File: golang/medium/78.子集.go
/*
 * @lc app=leetcode.cn id=78 lang=golang
 *
 * [78] 子集
 *
 * https://leetcode-cn.com/problems/subsets/description/
 *
 * algorithms
 * Medium (74.30%)
 * Likes:    349
 * Dislikes: 0
 * Total Accepted:    37.7K
 * Total Submissions: 50.4K
 * Testcase Example:  '[1,2,3]'
 *
 * 给定一组不含重复元素的整数数组 nums，返回该数组所有可能的子集（幂集）。
 *
 * 说明：解集不能包含重复的子集。
 *
 * 示例:
 *
 * 输入: nums = [1,2,3]
 * 输出:
 * [
 * ⁠ [3],
 * [1],
 * [2],
 * [1,2,3],
 * [1,3],
 * [2,3],
 * [1,2],
 * []
 * ]
 *
 */

func subsets(nums []int) [][]int {
	res := [][]int{}
	size := len(nums)
	if size == 0 {
		return res
	}
	res = append(res, []int{})
	for _, num := range nums {
		for _, ans := range res {
			t := []int{num}
			t = append(t, ans...)
			res = append(res, t)
		}
	}
	return res
}

File: labuladong/开始之前/backtracking/78.子集.py
#
# @lc app=leetcode.cn id=78 lang=python3
#
# [78] 子集
#

# @lc code=start
from typing import List
from copy import copy


class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = []
        now = []

        def backtrack(now, start):
            res.append(now)
            for i in range(start, len(nums)):
                now.append(nums[i])
                backtrack(copy(now), i + 1)
                now.pop()

        backtrack(copy(now), 0)
        return res

    def test(self):
        return self.subsets([1, 2, 3])


# @lc code=end

